File: test-suite/src/test/java/io/micronaut/mqtt/docs/consumer/acknowledge/type/AcknowledgeSpec.java
package io.micronaut.mqtt.docs.consumer.acknowledge.type;

import io.micronaut.context.ApplicationContext;
import io.micronaut.mqtt.test.AbstractMQTTTest;
import org.junit.jupiter.api.Test;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

public class AcknowledgeSpec implements AbstractMQTTTest {

    @Test
    void testAckingWithAcknowledgement() {
        ApplicationContext applicationContext = startContext();

// tag::producer[]
ProductClient productClient = applicationContext.getBean(ProductClient.class);
productClient.send("message body".getBytes());
productClient.send("message body".getBytes());
productClient.send("message body".getBytes());
productClient.send("message body".getBytes());
// end::producer[]

        ProductListener productListener = applicationContext.getBean(ProductListener.class);

        try {
            await().atMost(5, SECONDS).until(() ->
                    productListener.messageCount.get() == 4
            );
        } finally {
            applicationContext.close();
        }
    }
}

File: test-suite-python/src/test/python/micronaut/mqtt/docs/consumer/acknowledge/type/AcknowledgeSpec.py
from time import sleep
from typing import Annotated

from jakarta.inject import Inject
from micronaut.context.annotation import Property
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from .ProductClient import ProductClient
from .ProductListener import ProductListener


@Property(name="spec.name", value="AcknowledgeSpec")
@MicronautTest(environments=["mqtt"])
class AcknowledgeSpec:
    product_client: Annotated[ProductClient, Inject]
    product_listener: Annotated[ProductListener, Inject]

    @Test
    def test_acking_with_acknowledgement(self):
        # tag::producer[]
        self.product_client.send(b"message body")
        self.product_client.send(b"message body")
        self.product_client.send(b"message body")
        self.product_client.send(b"message body")
        # end::producer[]

        for _ in range(50):
            if self.product_listener.message_count == 4:
                break
            sleep(0.1)
        assert self.product_listener.message_count == 4
